Add support for boolean indexing

gvtulder opened the pull request:
This would be the easy way of adding support for boolean indexing, as suggested in #5192. It passes the simple tests, but it's not unlikely that there are edge cases that I didn't consider.

nouiz commented on theano/tensor/var.py:
```diff
@@ -460,30 +460,48 @@ def astype(self, dtype):
+            # normal integer indices.
+            if (isinstance(arg, (np.ndarray, theano.tensor.Variable)) and
+                    hasattr(arg, 'dtype') and hasattr(arg, 'nonzero') and
+                    arg.dtype == 'bool'):
```
you can simplify this condition:

    if isinstance(arg, (np.ndarray, theano.tensor.TensorVariable) and arg.dtype == 'bool':

gvtulder replied:
That's what I tried first, but that fails on this test:
```python
    assert_array_equal(numpy_n[:, mask], n[:, self.shared(mask)].eval())
```
because `theano.tensor.sharedvar.TensorSharedVariable` is not a `TensorVariable`. There is a similar problem `TensorConstant`, `GpuArraySharedVariable` et cetera. This was the least ugly solution I could think of that covers all those types.

nouiz replied:
ok. You could remove the hasattr(arg, 'dtype'), as currently all the var that have nonzero have dtype. But if you prefer to keep it to safeguard again possible futur that will probably not happen, you can keep it.

nouiz commented on theano/tensor/tests/test_subtensor.py:
```diff
@@ -362,6 +362,51 @@ def test_ellipsis(self):
+        mask = np.array([True, False, True])
+        assert_array_equal(numpy_n[0, mask], n[0, mask].eval())
+        assert_array_equal(numpy_n[:, mask], n[:, mask].eval())
+        assert_array_equal(numpy_n[:, mask], n[:, self.shared(mask)].eval())
```
Can you test with a partial slice too? Just in case.

nouiz commented on theano/tensor/tests/test_subtensor.py:
```diff
@@ -362,6 +362,51 @@ def test_ellipsis(self):
+        assert_array_equal(numpy_n[:, mask], n[:, mask].eval())
+        assert_array_equal(numpy_n[:, mask], n[:, self.shared(mask)].eval())
+
+        # indexing with a boolean array
```
array -> ndarray

nouiz commented on theano/tensor/tests/test_subtensor.py:
```diff
@@ -362,6 +362,51 @@ def test_ellipsis(self):
+        # for the GpuArray tests, self.shared will create GpuArray variables,
+        # but we also need to test the combination of GPU and normal TensorVariables
+        n_cpu = theano.shared(numpy_n)
+        assert_array_equal(numpy_n[mask], n[theano.shared(mask)].eval())
```
this won't test on the GPU. The GPU tests don't enable the GPU by default. They need to manually pass the mode. You can use self.mode as the mode to use to enable the GPU optimization to make the test do what you want.

But I don't think we need to test that. I would just remove that test. This is the normal workflow and should work without problems.

nouiz commented on theano/tensor/tests/test_subtensor.py:
```diff
@@ -362,6 +362,51 @@ def test_ellipsis(self):
+        self.assertRaises(TypeError, n.__getitem__, (True, False))
+        self.assertRaises(TypeError, n.__getitem__, ([True, False]))
+        self.assertRaises(TypeError, n.__getitem__, ([0, 1], [0, False]))
+        self.assertRaises(TypeError, n.__getitem__, ([0, 1], [0, theano.shared(True)]))
```
Also add a test where the mask is too big of too small. Just to make sure we raise the same error as numpy on CPU and GPU.

nouiz commented:
See my comment in the issue this address. Just in case you wan to do that version of the implementation. I'm good with both of them.

gvtulder commented:
Thanks for the comments. I'll have a look to see how difficult it is to move this to the GPU implementation of AdvancedSubtensor, so that the CPU version doesn't use `nonzero`.

The issue of checking the size of the mask is tricky. Right now, it doesn't do that: you can use a boolean mask that is larger than the tensor you're indexing, as long as there are no `True`s outside the range of your tensor. You can also pass a smaller mask. I agree that it's better to check the size, which would most likely require an `Assert`. The problem is that checking the dimensions gets a bit complicated when you allow for an ellipsis.

I don't fully understand how the ellipsis is handled at the moment: if I do this
```python
a = np.array(...) # some 2D array
x = T.tensor4()
x[a, ..., 0]
```
then it looks like the `__getitem__` method expands the ellipsis to `[a, slice(None), slice(None), 0]`, which is one dimension too many. Yet it still seems to work fine, so I have to look at this in a bit more detail.

nouiz commented:
For the size, I just checked and numpy currently just do a deprecation
warning for smaller mask and accept longer mask without warning if they are
only False. If True it raise an error. As you wrote.

So, can you just add test that it work? So we don't break it by accident in
the futur. If we break it volontarily, we can easily change the test to
request it to raise an error.

For elipsis. I don't understand why you expected one less dimensions. There
can be only 1 ellipsis. It get replaced by as many full slices as needed to
have the same number of argument then the inputs ndim. So 4 in your
example. Then the normal subtensor logic apply.

On Thu, Aug 3, 2017 at 10:29 AM Gijs van Tulder <notifications@github.com>
wrote:

> Thanks for the comments. I'll have a look to see how difficult it is to
> move this to the GPU implementation of AdvancedSubtensor, so that the CPU
> version doesn't use nonzero.
>
> The issue of checking the size of the mask is tricky. Right now, it
> doesn't do that: you can use a boolean mask that is larger than the tensor
> you're indexing, as long as there are no Trues outside the range of your
> tensor. You can also pass a smaller mask. I agree that it's better to check
> the size, which would most likely require an Assert. The problem is that
> checking the dimensions gets a bit complicated when you allow for an
> ellipsis.
>
> I don't fully understand how the ellipsis is handled at the moment: if I
> do this
>
> a = np.array(...) # some 2D array
> x = T.tensor4()
> x[a, ..., 0]
>
> then it looks like the __getitem__ method expands the ellipsis to [a,
> slice(None), slice(None), 0], which is one dimension too many. Yet it
> still seems to work fine, so I have to look at this in a bit more detail.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6207#issuecomment-319985758>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-zwPh0DviMQxxKKbuasI45H7dfrEks5sUdlSgaJpZM4OlbOZ>
> .
>

gvtulder commented:
Different sizes will [raise an error in numpy 0.13.0](https://docs.scipy.org/doc/numpy-1.13.0/release.html#boolean-indexing-changes). I think I'd prefer Theano to raise an error right now, instead of allowing it now and removing it again in the near future.

> For elipsis. I don't understand why you expected one less dimensions.

Sorry, that might have been my misunderstanding of how the indexing works. I expected that if `a` is 2D boolean ndarray and `x` is a 4D ndarray, then `x[a]` would be the same as `x[a, ...]`, but apparently it isn't. Now that I know that, I think it makes sense.

nouiz commented:
Honestly, I always forget what happen when you index with a tensor of indices/boolean. I don't understand why this is useful too. It probably just remove a reshape/dimshuffle. I never use it myself, I find it too confusing.

To make it raise an error for smaller size, it is easy if all the indexing is done in one op, like the AdvancedSubtensor class on the CPU. For the GPU, doing it in the graph would be complicated. We now have a GpuAdvancedIndexing op that only have python code. I think the perform could be modified to check the shapes and call nonzeros() on the index at the start and let the rest of the method unchanged. What do you think of that?

lamblin commented:
@nouiz It corresponds to selecting elements with a mask, rather than indices. See https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#boolean-array-indexing

gvtulder commented:
Something like this? The shape checks are now done in the `perform()` methods and `nonzero` is only used for the GPU version.

There's a little issue with the type of the error that is raised when the index dimension is larger than the dimension of the indexed array. In those cases, NumPy raises an `IndexError: too many indices for array`. Theano, currently, has two options: for simple cases, Subtensor [raises a `ValueError`](https://github.com/Theano/Theano/blob/61cb8c4/theano/tensor/subtensor.py#L474-L479), whereas for complex cases AdvancedSubtensor relies on NumPy to throw an `IndexError`. The new code would always raise an `IndexError`, but that breaks [these Subtensor tests](https://github.com/Theano/Theano/blob/61cb8c4/theano/tensor/tests/test_subtensor.py#L188-L195). What's the best way to solve this?

nouiz commented:
I think it make sence to move in the direction of being more like NumPy. I don't like the idea of having different error type for different indexing error. They are all indexing errors. So I would update the tests to expect IndexError.

nouiz commented on theano/tensor/var.py:
```diff
@@ -499,25 +503,22 @@ def includes_bool(args_el):
             args = list(args)
             args[ellipsis_at: ellipsis_at + 1] = (
-                [slice(None)] * (self.ndim - (len(args) - 1 - new_axes)))
+                [slice(None)] * (self.ndim - index_dim_count))
```
This change the current logic how we handle indices array that have more then 1 dimensions. As that what you want? What was the problem in how we handle them?

gvtulder replied:
Yes, that's on purpose, because the boolean arrays are treated differently from integer arrays (see my other comment). An integer array counts for 1 dimension no matter what shape it is, but a boolean array counts for ndim dimensions. The loop a few lines before this does that counting.

nouiz replied:
right

nouiz commented on theano/tensor/var.py:
```diff
@@ -533,7 +534,12 @@ def includes_bool(args_el):
         for i, arg in enumerate(args):
             try:
-                if arg is not np.newaxis:
+                if (isinstance(arg, (np.ndarray, theano.tensor.Variable)) and
```
Why not modify Subtensor.convert() to also raise AdvancedIndexingError? This would make sure that elsewhere where we call convert(), if we forgot to update the code, we will end up using advanced indexing.

gvtulder replied:
I moved the test to Subtensor.convert() now.

It still requires a test here, though: if there is a boolean mask, it should be given to AdvancedSubtensor and not AdvancedSubtensor1, even if there is only one mask.

gvtulder commented on theano/gpuarray/subtensor.py:
```diff
@@ -481,6 +481,37 @@ def c_code_cache_version(self):
+            # skip, does not count as an input dimension
+            out_idx_list.append(index)
+        elif isinstance(index, np.ndarray) and index.dtype == 'bool':
+            for i in xrange(index.ndim):
```
I find it confusing as well, but I think what numpy does is this:
- for multi-dimensional integer arrays, it uses the values as indices for a single dimension (and it reshapes the output in that dimension to have the same shape as the index array);
- for boolean arrays, it uses them as a mask, so that if you pass a boolean matrix as the first indexing element, this will be applied to the first two dimensions of the indexed array.

Example of how I think it works:
```python
x = np.arange(16).reshape(2, 2, 2, 2)

# integer index array
a = np.zeros((2, 2), int)
x[a, 0, 0, 0]  # returns [[0, 0], [0, 0]]
# the indices in matrix a applies only to the first dimension of x

# bool mask array
b = np.array([[True, False], [False, True]])
x[b, 0, 0]  # returns [0, 12]
# the first dimension of mask b applies to the first dimension of x,
# the second dimension of b to the second dimension of x,
# so only the third and fourth remain to be indexed
```
For integer arrays, the shape of the index array doesn't matter, because they're used as indices for a single dimension. For boolean arrays, the shape must be the same as that in the corresponding dimensions of the input.

gvtulder commented on theano/tensor/tests/test_subtensor.py:
```diff
@@ -382,41 +391,66 @@ def test_boolean(self):
+        assert_array_equal(numpy_inc_subtensor(numpy_n, [0, mask], 1),
+                           inc_subtensor(n[0, mask], 1).eval())
+        assert_array_equal(numpy_inc_subtensor(numpy_n, [Ellipsis, mask], 1),
+                           inc_subtensor(n[:, mask], 1).eval())
```
Good catch, that's supposed to be a `slice(None)`.

nouiz commented:
I got a lenghy IRL discussion with @lamblin Friday. Here is his comment that I agree with.

- If you remember, we go some bugs due to some optimization not checking the set_instead_of_inc props of IncSubtensor. Reusing the same old op for boolean indeixing  and advanced indexing risk that. We don't like that. So we would prefer to have a new AdvancedBooleanIndexing op. There could be a base class that both AdvancedBooleanIndexing and AdvancedIndexing to share much of the code.
- I don't understand why all tests pass when you didn't update the make_node and infer_shape of  AdvancedIndexing. Check your own example, one return 2d and one return 1d. Is there something I'm missing of tests are missing?

@lamblin don't care if we don't return now shapes errors as the new numpy 1.13. He don't see problem of using nonzeros in the graph for now. Theano already have some cases where we don't returns errors.

We try to make a Theano beta this week and rc next week. So we should merge new feature ideally this week. If you don't have time to finish this PR, can you make a new PR with the nonzero to have this in Theano 0.10?

We won't have time to finish this ourself as a few people are in vacation.

gvtulder commented:
> If you remember, we go some bugs due to some optimization not checking the set_instead_of_inc props of IncSubtensor. Reusing the same old op for boolean indeixing and advanced indexing risk that. We don't like that. So we would prefer to have a new AdvancedBooleanIndexing op.

I don't fully understand the reason, but I've now added a new `AdvancedBooleanSubtensor` op that shares a `BaseAdvancedSubtensor` class with `AdvancedSubtensor`. Is this more or less what you meant? (I don't know if it's necessary to have two classes for the gpu version as well?)

> I don't understand why all tests pass when you didn't update the make_node and infer_shape of AdvancedIndexing. Check your own example, one return 2d and one return 1d. Is there something I'm missing of tests are missing?

- For `infer_shape`, you're right: that wasn't tested. The `infer_shape` of `AdvancedSubtensor` doesn't do much anyway, but the special case that is implemented did give a problem with boolean indexing.

- `make_node` doesn't need to be changed, because it gets its shape information from `adv_index_broadcastable_pattern`, and [I did have to change that](https://github.com/Theano/Theano/pull/6207/files#diff-b8180324f78f23b3f0df24954e3d1e54R2104). That should solve the 2D/1D problem.

nouiz commented on theano/tensor/subtensor.py:
```diff
@@ -40,6 +41,16 @@ def __init__(self, *args):
+
+    """
+
+    def __init__(self, *args):
```
You don't need this init method. Just use "pass" as the content of this class for python syntax.

nouiz commented on theano/tensor/subtensor.py:
```diff
@@ -2170,11 +2215,71 @@ def grad(self, inputs, grads):
         return [advanced_inc_subtensor(theano.tensor.zeros_like(x), gz,
```
remove grad from the base class.

gvtulder commented:
Fixed.

nouiz commented on theano/gpuarray/opt.py:
```diff
@@ -1076,6 +1076,13 @@ def local_gpua_advanced_subtensor(op, context_name, inputs, outputs):
+@op_lifter([tensor.AdvancedBooleanSubtensor])
+@register_opt2([tensor.AdvancedBooleanSubtensor], 'fast_compile')
+def local_gpua_advanced_boolean_subtensor(op, context_name, inputs, outputs):
+    return GpuAdvancedSubtensor()
```
All the PR look good, but I have one thing that make me hesitant. On the GPU, you have just one op that also does the boolean indexing. This is different then what we do. I'm not opposed to it, but I would prefer that someone else is also ok about this before merging.

Also, having just one class named GpuAdvancedBooleanSubtensor would be more consistent to what the CPU does. What do you think about this?

gvtulder commented:
I've now split the GpuAdvancedSubtensor into a boolean and non-boolean class. I have mixed feelings about this: it's nice that it mirrors the CPU structure more closely, but it's also a bit weird to have two classes that do exactly the same thing. The same holds for the CPU implementation, where the non-boolean op is just a restricted case of the boolean op. Since the CPU version has two classes, I guess the neatest solution is to map it to a similar structure on the GPU.

Note that I've also added the GpuAdvancedBooleanIncSubtensor op and a disabled optimization, but that op still has bugs. The tests fail if you enable the optimization. I think it's useful to have the optimization so that it won't be forgotten in the IncSubtensor repairs.

nouiz commented:
I'm also not sure what to do for the same reason. I would like to discuss
this with @abergeron or lamblin before merging for this reason.

I don't like the idea of having 2 ops that does exactly the same thing on
the GPU. If we go in that direction the nonbool op should need to raise an
error like on the CPU if bool are used with it.

What I think is a good middleground would be to just implement the bool
version on the GPU. So that keep the corresponding with the GPU and it make
it clear in the tests and to the developer that it should also support the
bool case. (then we don't implement the nonbool op on the GPU). For the non
bool optimization, the would continue to be done on the CPU as before.

What do you think of that?

On Thu, Aug 10, 2017 at 6:27 AM Gijs van Tulder <notifications@github.com>
wrote:

> I've now split the GpuAdvancedSubtensor into a boolean and non-boolean
> class. I have mixed feelings about this: it's nice that it mirrors the CPU
> structure more closely, but it's also a bit weird to have two classes that
> do exactly the same thing. The same holds for the CPU implementation, where
> the non-boolean op is just a restricted case of the boolean op. Since the
> CPU version has two classes, I guess the neatest solution is to map it to a
> similar structure on the GPU.
>
> Note that I've also added the GpuAdvancedBooleanIncSubtensor op and a
> disabled optimization, but that op still has bugs. The tests fail if you
> enable the optimization. I think it's useful to have the optimization so
> that it won't be forgotten in the IncSubtensor repairs.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6207#issuecomment-321513715>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-83Llzc0Hm4s4kliZQ6riqztpcsyks5sWtsKgaJpZM4OlbOZ>
> .
>

gvtulder commented:
I see that a new error has appeared in Jenkins. Something to do with 16-bit floats. I'll have to look into that.

> I don't like the idea of having 2 ops that does exactly the same thing on
the GPU. If we go in that direction the nonbool op should need to raise an
error like on the CPU if bool are used with it.

Yes, that's what I had done now, both on the CPU and on the GPU. The nonbool versions of the ops are exactly the same as the bool versions, except that the nonbool ops run a `check_and_reject_bool(index)` to throw an error if there are any boolean indices.

> What I think is a good middleground would be to just implement the bool
version on the GPU. So that keep the corresponding with the GPU and it make
it clear in the tests and to the developer that it should also support the
bool case. (then we don't implement the nonbool op on the GPU). For the non
bool optimization, the would continue to be done on the CPU as before.

I'm not sure what you mean by this. Do you mean:
* CPU: keep `AdvancedSubtensor`, `AdvancedBooleanSubtensor`, `AdvancedIncSubtensor` and `AdvancedBooleanIncSubtensor`
* GPU: rename `GpuAdvancedSubtensor` and `GpuAdvancedIncSubtensor` to `GpuAdvancedBooleanSubtensor` and `GpuAdvancedBooleanIncSubtensor`

In that case I think I'd prefer it to be the way it is now: each CPU op maps to a `Gpu*` equivalent. Wouldn't removing/renaming existing ops cause problems?

nouiz commented on theano/gpuarray/subtensor.py:
```diff
@@ -634,21 +622,35 @@ def perform(self, node, inputs, out_):
+        return gof.Apply(self, [x] + rval.inputs[1:], [otype()])
+
 
+class BaseGpuAdvancedIncSubtensor(object):
```
This look great. Can you make the non bool op raise an error if it receive a bool as input as on the CPU? Maybe just an attribute support_bool=True/False in each sub class and have the perform reuse that?

It is not great to have the CPU and GPU op not behave the same. It caused subtle user problem in the past (code that worked only on GPU, but not on CPU)

After that we should be ok to merge.

gvtulder replied:
This is already checked, but it is done in `AdvancedSubtensor.make_node`. The `GpuAdvancedSubtensor.make_node` calls that function, so it's automatically checked and already raises `IndexError`s for boolean indices.

But it is a bit hidden and implicit. Two options to improve that:
* Repeat the check in `GpuAdvancedSubtensor.make_node` (by calling `tensor.subtensor.check_and_reject_bool`). Then the inputs are checked twice.
* Add a comment to `GpuAdvancedSubtensor.make_node` explaining that the checks are done in the CPU op's `make_node`.

nouiz commented on theano/gpuarray/subtensor.py:
```diff
@@ -481,24 +481,46 @@ def c_code_cache_version(self):
 
 
-class GpuAdvancedSubtensor(HideC, tensor.AdvancedSubtensor):
+def check_and_convert_boolean_masks(input, idx_list):
```
Just an idea, if you make an attribute to the op as said in the other comment, mostly, you could have only one implementation of this function in the CPU code that have a parameter to raise or convert.

gvtulder replied:
The checks are done at `make_node` time, not at runtime in `perform`, so there is only one implementation. I thought it wouldn't do much harm to have the conversion in the `perform` for the non-bool op as well, because it just doesn't change anything and having a one-size-fits-all method keeps things simple.

nouiz replied:
ok. thanks merging.
